File: backend/joblisting.py
import uuid

from flask import current_app, jsonify, Blueprint, render_template, request
from flask_login import login_required, current_user

from backend.utils.prepare_document import prepare_document

jobListing = Blueprint('jobListing', __name__)


@jobListing.route("/createJobListing", methods=["GET", "POST"])
@login_required
def createJobListing():
    if not current_user.is_authenticated:
        return jsonify({"Error": "You are not logged in"}), 403

    if request.method == 'POST':
        jobListingData = request.json.get('formData')

        if not jobListingData:
            return jsonify({"Error": "No job listing provided"}), 403

        try:
            _id = uuid.uuid4().hex

            if int(jobListingData['work_hours']) < 1:
                return jsonify({"Error": "Work hours must be greater than 0"}), 403

            variables = {
                "_id": _id,
                "user_id": current_user.get_id(),

                "title": jobListingData['title'],
                "description": jobListingData['description'],
                "category": jobListingData['options'].get('category')[0],

                # optional
                "additional_information": jobListingData['additional_information'],
                "payment_type": jobListingData['options'].get('payment_type')[0],
                "payment_amount": int(jobListingData['payment']),
                "start_date": jobListingData['start_date'],
                "days": jobListingData['days'],
                "hours": int(jobListingData['work_hours']),
            }

            document = prepare_document('jobListings', variables)

            current_app.config['DB'].Insert("jobListings", document)

            return jsonify({"Success": "Job listing created successfully"}), 200
        except Exception as e:
            print("Error:", e)
            return jsonify({"Error": "There has been an error, please try again later"}), 403

    return {}, 200


@jobListing.route("/getJobListingById", methods=["GET"])
def getJobListingById():
    if request.method == 'GET':
        job_id = request.args.get('currentJobIdFromSearch')

        if not job_id:
            return jsonify({"Error": "No job id provided"}), 403

        job = current_app.config['DB'].Find(
            "jobListings", {"_id": job_id})

        if not job:
            return jsonify({"Error": "No job found"}), 403

        job['location'] = job.get(
            'location', None) or "Remote"
        job['hours'] = job.get('hours', None) or 0

        newJob = {
            "id": job['_id'],
            "user_id": job['user_id'],
            "title": job['title'],
            "description": job['description'],
            "category": job['category'],
            "additional_information": job['additional_information'],

            "tags": {
                "payment_type": job['payment_type'],
                "payment_amount": str(job['payment_amount']),
                "start_date": job['start_date'],
                "days": job['days'],
                "category": job['category'],
                "hours": job['hours'],
            },

            "posted_at": job['created_at'],
            "location": job['location'],
        }

        return jsonify(newJob), 200

    return {}, 200


@jobListing.route("/getJobListings", methods=["GET"])
def getJobListings():
    if request.method == 'GET':
        jobListings = list(current_app.config['DB'].FindAll("jobListings", {}))

        if not jobListings:
            return jsonify({"Error": "No job listings found"}), 403

        newJobListings = []

        for job in jobListings:
            job['location'] = job.get(
                'location', None) or "Remote"
            job['hours'] = job.get('hours', None) or 0

            newJob = {
                "id": job['_id'],
                "user_id": job['user_id'],
                "title": job['title'],
                "description": job['description'],
                "category": job['category'],
                "additional_information": job['additional_information'],

                "tags": {
                    "payment_type": job['payment_type'],
                    "payment_amount": str(job['payment_amount']),
                    "start_date": job['start_date'],
                    "days": job['days'],
                    "category": job['category'],
                    "hours": job['hours'],
                },

                "posted_at": job['created_at'],
                "location": job['location'],
            },

            newJobListings.append(newJob)

        return jsonify(newJobListings), 200


@jobListing.route("/getUserByIdForJobListing", methods=["GET"])
def getUserByIdForJobListing():
    if request.method == 'GET':
        user_id = request.args.get('user_id')

        if not user_id:
            return jsonify({"Error": "No user id provided"}), 403

        user = current_app.config['DB'].Find(
            "users", {"_id": user_id})

        if not user:
            return jsonify({"Error": "No user found"}), 403

        newUserData = {
            "id": user['_id'],
            "name": user['first_name'] + " " + user['last_name'],
        }

        return jsonify(newUserData), 200

    return {}, 200

File: server/controllers/joblisting.js
const express = require('express');
const crypto = require('crypto');
const asyncHandler = require('express-async-handler');

const { clerkClient } = require('@clerk/clerk-sdk-node');

const db = require('../services/db');
const prepareDocument = require('../utils/prepare_document');
const { formatDate } = require('../utils/utils');

const { NewJobListingSchema } = require('../schemas/joblisting');

exports.getJobListings = asyncHandler(async (req, res) => {
    const jobListings = await db.collection('jobListings').find({}).toArray();

    if (!jobListings) {
        return res.status(500).json({
            message: 'Failed to get job listings'
        });
    }

    res.status(200).json({
        data: jobListings
    });
});

exports.getJobListingsByUserId = asyncHandler(async (req, res) => {
    const userId = req.params.id;

    const jobListings = await db.collection('jobListings').find({ user_id: userId }).toArray();

    if (!jobListings) {
        return res.status(500).json({
            message: 'Failed to get job listings'
        });
    }

    res.status(200).json({
        data: jobListings
    });
});

exports.getJobListingsFromJobId = asyncHandler(async (req, res) => {
    const jobId = req.params.id;
    const jobListing = await db.collection('jobListings').findOne({ _id: jobId });

    if (!jobListing) {
        return res.status(500).json({
            message: 'Failed to get job listing'
        });
    }

    let clerkUser;
    try {
        clerkUser = await clerkClient.users.getUser(jobListing.user_id);
    } catch (error) {
        const deleteListing = await db.collection('jobListings').deleteOne({
            _id: jobId
        });

        if (!deleteListing) {
            return res.status(500).json({
                message: 'Failed to delete job listing'
            });
        }

        return res.status(400).json({
            error: 'UserNotFound'
        });
    }

    jobListing.person = {
        id: clerkUser.id,
        firstName: clerkUser.firstName,
        lastName: clerkUser.lastName,
        fullName: clerkUser.fullName,
    }

    res.status(200).json({
        data: jobListing
    });
});

exports.getJobListingUserFromId = asyncHandler(async (req, res) => {
    const user = req.params.id;
    let clerkUser;

    try {
        clerkUser = await clerkClient.users.getUser(user);
    } catch (error) {
        const deleteListing = await db.collection('jobListings').deleteOne({
            user_id: user
        });

        if (!deleteListing) {
            return res.status(500).json({
                message: 'Failed to delete job listing'
            });
        }

        return res.status(400).json({
            error: 'UserNotFound'
        });
    }

    const userData = {
        id: clerkUser.id,
        firstName: clerkUser.firstName,
        lastName: clerkUser.lastName,
        fullName: clerkUser.fullName,
    }

    res.status(200).json({
        data: userData
    });

});

exports.createJobListing = asyncHandler(async (req, res) => {
    const data = req.body;
    const form = data.formData;

    const _id = crypto.randomBytes(16).toString('hex');
    const formatedDate = formatDate(new Date());
    const newJobListing = {
        _id: _id,
        user_id: form.user_id,

        title: form.title,
        description: form.description,
        category: form.options.category[0],

        additional_information: form.additional_information,
        payment_type: form.options.payment_type[0],
        payment_amount: parseInt(form.payment),
        start_date: form.start_date,
        days: form.days,
        hours: parseInt(form.work_hours),
        location: form.location || 'N/A',
        posted_at: formatedDate
    }

    const document = prepareDocument(NewJobListingSchema, newJobListing);
    const jobListingInsert = await db.collection('jobListings').insertOne(document)

    if (!jobListingInsert) {
        return res.status(500).json({
            message: 'Failed to create job listing'
        });
    }

    res.status(200).json({
        message: 'POST job listing'
    });
});

exports.deleteJobListing = asyncHandler(async (req, res) => {
    const jobId = req.params.id;

    if (!jobId) {
        return res.status(400).json({
            message: 'Missing job id'
        });
    }

    const deleteListing = await db.collection('jobListings').deleteOne({
        _id: jobId
    });

    if (!deleteListing) {
        return res.status(500).json({
            message: 'Failed to delete job listing'
        });
    }

    res.status(200).json({
        message: 'DELETE job listing'
    });
})